### ISSUE TITLE
A few ideas for enhancements

### ISSUE THREAD

**Issue by Pastim** [medoc92/upplay#4]
Hi,

A few ideas for enhancements:

1) Support searches for items - look at bubbleupnp for an example.

2) Support random play for a set of items in a container (not just the whole list).  This is not the same as shuffle of a playlist.  The idea is to find all the (unique) items in a container, select  a random n and send these to the playlist.  When the playlist is running out, add another n.  This allows one to sit all day listening to things you don't normally select, and find unusual contrasts between consecutive items.  The normal playlist shuffle facility can't do this this one would have to send thousands of items to the playlist.

3) Extend (2) to be able to play random 'groups' of items.  I don't know if this is possible for a control point, or whether it may only be possible using minimserver.  The idea is that a set of items may comprise a 'Work' (eg a symphony) or a minimserver 'Group' (which might be one act of an opera), or a whole Album, and one would like to play all the items in the group in their correct order, and then move on to another random group.  I don't know whether the Control Point can even 'see' the meaning of such groups, so it may be that this isn't possible.

**Comment by medoc92** [comment]
Hi, 

I'm getting back to the Playing Random 'works' thing (I know, long time...). I am trying to get the Minim "group" feature to work, with no success at the moment. I don't strictly need it, because, as seen from the player a "group" is just another container, but I'm baffled. Before I go the Minim support list, could you please tell me what tag you use for grouping ? I tried with id3v2 TIT1 - CONTENTGROUP in the MinimServer tag mapping table - but no containers appear...

**Comment by Pastim** [comment]
MinimServer uses the 'Group' tag.  I have previously used a different tag to group 'movements' together, called 'Work', but since MinimServer uses Group I have duplicated Work to Group on my flacs.  As you say, Group provides a container, whereas my 'Work' can be used for indexing and browsing.  

Sequential sets of tracks on an album with the same Group tag automatically appear as one container on my system using upplay.  If the same Work appears on an Album (which does happen occasionally) I add a suffix to the Group tag to distinguish them.

See http://minimserver.com/ug-browsing.html#Groups and compositions.

**Comment by medoc92** [comment]
Thanks, what I needed to do was establish an aliasTag from CONTENTGROUP to GROUP

**Comment by medoc92** [comment]
Hi,

I've pushed a first version of the "random play" feature to the repository. I sort of remember that you have no issue with building upplay, but, else, I can push a binary version somewhere (you'll need to remind me of what system you use).

The feature is accessed from the right-click menu over any container in the directory browser.

I prefer to not describe it too much, I'd rather put you in the situation of the normal user who did not read the manual. What you'll find difficult to understand or discover will need changing. It's quite basic in any case.

But of course, don't lose time, and ask me if something is not clear.

The menu entry is "play by albums", but, as it seems that Minim sets the album name to the group name, this should hopefully do what we want. Else, I've other ideas about how to segment the list.

Let me know how this works for you !

**Comment by Pastim** [comment]
I'm having some compilation problems (xubuntu 15.10).  Do I need to update libupnpp as well?  I normally use qmake, followed by sudo checkinstall (which does a proper reversible install).  I have been using the master zip each time, having been slightly unclear what I get using git pull, particularly when version numbers change.

On my system, my Minim album names are not the group names.  That may be an option somewhere, but my albums are definitely albums in the true sense of the word.

**Comment by medoc92** [comment]
I can easily build some packages tomorrow. What is your machine architecture (uname -m) ?

When you display a list of tracks in the upplay directory browser, and the tracks are part of a group, what do you see in the "album" column ? I am seeing the group name (contentgroup/tit1 in my case).

**Comment by Pastim** [comment]
OK - compilation sorted.  I needed to update libupnpp first.

The result of 'play by albums' (please don't call it that!) is not quite as expected.  If I go to a composer, and request this, I get all the versions of one work (group), from all albums, with all the 1st movements 1st, then 2nd, and so on.  If I clear the playlist, it goes on to another work.   

What the random play should do (in my humble opinion) is play single random instances of complete works, if you get what I mean,

It would make things easier to follow if the playlist included up to n works as well (but bear in mind a work can have 50+ tracks).

However, it's definitely getting there!

**Comment by Pastim** [comment]
Our posts overlapped.  

In response to your query, surely what you see is the track names - album on which the tracks reside - Group - time.  Or am I confused (quite possible)

**Comment by medoc92** [comment]
About the groups mixup, obviously it's not what I intended, I did not see the issue because I have a single album with "works" which I created for trying this, so works are unique, and there is no issue of multiple track X.

I need to properly tag a couple of albums so that I can see how things appear under the different Minimserver views, and how/if  I can sort things out.

Currently the grouping is purely done in the "album" column seen in the track listings (the one before the time). If there is a group, it appears that minimserver sets its value in there in place of the album tag found in the file.

Clearly, it's not enough to group things only from this value (because, as you saw, this will concatenate tracks from multiple albums into the same group).

More thinking is needed. If you have any idea about the question, don't hesitate :)

**Comment by Pastim** [comment]
I have 4 columns, not 3.  I'm puzzled by your definition of 'album' column.  I see:
1 - track name
2 - album name
3 - group name
4 - time

This is true on the sample you posted. 4 columns not 3, or am losing my marbles?

**Comment by Pastim** [comment]
Sorry, yes, I have lost my marbles.  I was reading your artists as an album name. Sorry.

If you do a search by Album, you get Album names, not Group names.  I therefore think minimserver is suppling group instead of album for that column where a group exists.  But it definitely isn't minimserver's Album name, which still exists.

So it is surely 'random play by group'?

And thanks for getting this far.  I think it's very close.

**Comment by Pastim** [comment]
If I play random tracks I get a load of tracks in the playlist, and can move around or skip some.  Would it possible to have a load of Groups (works) in the playlist instead of just one?

Do you depend on minimserver for this?  Could you do the same with other servers and a user-specified tag, or is that impossible?  I don't mind, because minimserver suits me very well indeed, but I wonder if other people might prefer other servers.

And apologies again for my stupid confusion.  I blame age, and doing too many things at once........

**Comment by medoc92** [comment]
Please don't apologize, the whole business is extremely confusing, and I have trouble navigating it too.

Actually, the problem is not even obvious for a human. If the same conductor records the same work twice the same day (same artists, different sessions), the only way you know what track goes with what is that they're consecutive on the CD (the sound engineer probably has a session identifier or something, but this does not come to us). The ISRC would be nice to have, but we don't have it, and even then, it might not be easy to use it to group things.

More thinking is in order. If any idea comes to you, I'm very receptive :)

**Comment by Pastim** [comment]
A combination of Album and Group should identify one instance of a Work.  I do this on another system (LMS - squeezebox) and it works well. There is one thing that has to be done when tagging to take care of the possibility you raise.  

If there is more than one version of a Work on an Album, you have to give them different Group names. I do this using a suffix that I add to the Group name when this happens.  If you have 2 Works with the same Group name, with some other tracks in between them, minimserver complains (see the scan log) - all tracks in a Group on an Album must be sequential.  

Note that I have many multi-disc Albums, where the discs are also numbered within the Album (using discnumber and disctotal tags).

**Comment by medoc92** [comment]
Don't forget that I don't have the luxury of finding values for Album *and* Group in the track data, it's only one upnp:album column where I will find an identifier which may match an Album or a Group, depending on how the Directory Server feels like. 

So, except if you have a better idea, I think that "play by albums" is actually not such a bad name, because this is what most users will see. Knowledgeable people will know (after reading a bit of doc) that an album may actually be a group, depending on the server and the tagging.

I have found a way to avoid group mixup though, by using the path in the tree. For minim this only works well for some parts of the tree, but I'm hoping  that this should be workable. For example under Composer/Bach, you should start random play on the Albums section (some other sections could work too, did not think of everything). If you start random play at the "Bach" level, I don't see any way to make things work, because the same track will be found under multiple paths, and I can't sort my chicken.

Otoh, I can find no way to do something which I would find interesting, which would be to play all the versions of a given work. The path which we would need in Minim would be something like Composer/Bach/Groups/brandenburg1/albums, but it does not exist. The groups appear under "items", but the same group name will be duplicated if there are several instances on different discs, instead of being at the head of an album hierarchy. And there is also the issue that you are giving different group names to the same composition.

In any case, I've pushed the new version, it should be an improvement if you use it as shown above(start on Albums), and a step for more thinking.

**Comment by Pastim** [comment]
Please call it "Play by Group", please...  It is randomly playing Groups, not randomly playing Albums.  If the name stays that way I'll remain very confused.  The fact that the Group Name is appearing in the Album column is, to my way of thinking, an internal technical problem, not the user's.  I don't want to know about that.  Albums are still Albums.  

I don't understand your point about the path, but I'll test it and see if it does what I expect or not.

To play all versions of a work (ignoring cases where there are multiples on an album), is easy (I think).  I expose the Group name on my index (in the IndexTags).  So I can go to Composer, then Group and play all of them either explicitly, or randomly using your need option.  In fact, my Group is derived from my Work tag, so I can also use my Work tag to do the same thing, even when there are multiples on an Album.

OK?

**Comment by medoc92** [comment]
No problem for the name, I'll change it. Not too sure about Groups and IndexTags, I would have to look at bit more at the Minim doc, maybe I'll try it, or maybe I won't need to if you tell me things work reasonably...

**Comment by Pastim** [comment]
Thanks for the name change!

With this version I randomly played one Group (selected from the Index), but the tracks were all mixed up from different versions (all on different albums) and not in sequence.  Not knowing what's being presented to upplay I don't know what to suggest (I have no idea what the protocol looks like).  Do you get a physical folder path for the tracks as well? I guess versions of a work would generally be in different folders.

I then tried to randomly play all of a particular genre and it then got stuck and wouldn't play any more.  I restarted upplay and tried again but once again it got stuck not accepting any more 'play' commands. Then it woke up.  Maybe it just takes a while to get a set to play?  Where is the randomisation taking place?  On the server or upplay client?  I don't see much activity on either.

**Comment by Pastim** [comment]
More on getting 'stuck'. Sometimes I get no response to a random play request, even after many minutes.  When I stop and restart and do the same request again, it may work, but give only a few Groups.  Something's not quite right somewhere.  Is there a log I can look at?

**Comment by medoc92** [comment]
About the path: there is not enough information inside the track items to do what we want. So I have to use the position of the track in the tree presented by MinimServer. Basically, I think that things should work at least  if there is a directory "Albums" in the current path, because this should hopefully ensure that the tracks don't get mixed up.

About playing from Genre, I think that walking a whole tree like this can take a lot of time, getting a bit more deeply into the tree might improve things. I'll see to adding a busy indicator later.

**Comment by Pastim** [comment]
I'm struggling to see how I can put Album in my search path for random play, since much of the idea (to me) is to play from different albums.  I might be able to offer an idea if I knew what data you were being given.

There's definitely more to getting stuck than a time delay.  I can wait forever, or restart, click the same thing again, and it seems to work quite quickly.  It's as if the completion of the process is sometimes being missed.  Is that possible?

**Comment by medoc92** [comment]
About getting stuck, yes, everything is possible at this point, I'll look into it.

About the albums, maybe it would help if I was using the same MInimServer configuration which you are using, could you share your config file ?

Also tell me from what point in the tree exactly you are trying to make this work, so that I can see why it does not and if it is possible to fix it.

**Comment by Pastim** [comment]
I'm still doing 5 things at once, so forgive me if I'm muddled.....

I can't attach anything here (not enough permissions), so the config file will come by email.

I can go anywhere high up in my tree, such as Genre, and the picking a Genre for Random play, even when there are only a few works/groups of that genre.

I even tried Classical (which is most of my library).  I expected it to be slow.  Nothing happened.  So eventually I stopped, restarted, and picked another track to play, and lo and behold it started to play random classical.

**Comment by medoc92** [comment]
I'm going to try things with your conf, and also look into why it sometimes gets lost while looking for tracks.

Just one more data point from your side: if you go to Composers/Some Composer/ and then right-click XX Albums and chose play by random albums, does it do the right thing ?

**Comment by Pastim** [comment]
Not quite.  I tried one Composer who has a few works of which I have 2 versions (on different albums).  It queues up both copies of the work in one go on the playlist, whereas I'd hope for only one, but this time  the tracks are in the correct order on each version of the work.  I have some works with many versions, and I'd have to skip over a load of tracks to get to the next random work.

**Comment by medoc92** [comment]
I pushed an update which should breakup the list more instead of sending multiple groups. 

By the way, you can skip a  list while keeping up random play by emptying the playlist (the broom top right above the list). This skips to the next group, faster than track by track (at least with upmpdcli, I'm a bit worried that this "feature" might be renderer-dependant).

I seem to have relatively good success for playing multiple versions of a group by going to, for example Composer/jsbach/Group/Some Concerto  and then starting random play by groups on the "nn  albums" entry.

I think that the player gets stuck when you start random play without having stopped it explicitely before, this should be a simple bug, I'll look into it next.

**Comment by Pastim** [comment]
Providing you select the random play on the Albums it does indeed return one group.  This works on several different selections so far, so that's very good progress, and already more flexible than I have from my other music server.

Using the broom also works to skip the group, as you say - I should have thought of that.

I still get stuck sometimes.  As far as I can see I can't start a random play unless the previous one has stopped, or are you saying that minimserver is still on an unfinished set (even if upplay has been restarted), so I have to restart, ask for another selection, and then stop to get back into sync?

**Comment by Pastim** [comment]
Just to say this is already brilliant.  Many of my evenings are spent reading and listening to random works.  This extension of upplay means I have even more flexibility than before.  I'll do some serious listening over the next few evenings, even if it does still get stuck sometimes (like right now...).

**Comment by medoc92** [comment]
I'm still not completely sure of what gets into trouble. There was a case which was certain, and now fixed, that if the random playing went to its end (signalled by an empty playlist), and you tried to restart it without using 'stop', you could not then see the popup any more. But there might be other possibilities.

You are supposedly allowed to start a new random sequence while one is already playing, the new one should just replace the old one.

I'll try to get it in trouble myself, but if you can narrow the conditions to reproduce the problem this will definitely be of help.

About using the broom: this was actually a good surprise for me, so you could hardly guess....

**Comment by Pastim** [comment]
If already playing random groups (thanks...) you can't request another one - the option isn't available on my system.

I'll try to pin down the problem later.

**Comment by medoc92** [comment]
What do you mean by 'not available' ? does the popup not show up? option greyed out ? Does nothing?

What I'm seeing is that starting this in the wrong place in the Minim tree can take literally forever, I need to implement a busy dialog and cancel button. But once started, I do get the normal popup, with the option to start again.

But I'll let you listen to a little music, this can be sorted out later ...

**Comment by Pastim** [comment]
Maybe I'm only seeing the pop up not being available (doesn't show at all) when it is 'waiting forever'.  Anyhow, I'll get to some listening and testing a bit now.

One passing thought was that if having 'Album' in the tree is essential, could you somehow enforce that yourself by doing an extra initial query to get to the Album count and then use that automatically?

**Comment by Pastim** [comment]
I can't work out when it sometimes gets stuck.  I'll keep looking.

Meanwhile a nice-to-have would be to visually expand the album art when hovering over it, or clicking,  The reason is that the album name doesn't show anywhere else when playing random groups, only the group name.  I've been playing a random selection of my highest-rated groups (strictly groups containing one or more highest rated tracks) for some hours, and it works well.

**Comment by medoc92** [comment]
About the Album thing: the reason this works well on the tree branches containing an Album listing is that I can use the information for sorting the tracks. The tracks themselves have no Album info (it's replaced by the Group value). But I have an other source information which is the location of the track in the tree. If there is an album entry in the path, sorting and splitting the list of track entries on path + Work value will compose consistent Work groups.

If there is no album, there may be nothing to differentiate tracks from the same group (as there is no album info in the track itself, the info is just not there). If we're lucky, there may be something else to help, like different Artist info if this is in the path, but having the Album is the best guarantee.

Using the artist info from the tracks to discriminate is probably not a good idea, both because the same artist could appear on separate recordings, and because artists could change between movements of the same group.

Album is our best chance, because, even if the same artists record several versions of the same work on one disc, the track numbers are bound to be different, so we're good.

All this would be very difficult to make automatic, because there is nothing to say that the "Album" listing will be named "nn Albums", minim allows to change this for example, and it might be totally different with another server. So, expert users will have to read the doc (when I've written it...), and naive users probably have no groups anyway, so the random play is going to play albums, which they'll hopefully find very nice.

Expanding the album art does look like a very good idea, this is actually the only element with album information while playing random groups (because I don't really interpret the path in the tree, just use it, I could not extract the Album name from it)

I pushed a new version with a cancel dialog when a directory walk takes more than 1 S

**Comment by Pastim** [comment]
I'll try the new version tonight.

I don't really understand the Albums issue.   Further, I still don't understand why you write 'random play is going to play albums', since you aren't playing albums, you are playing groups (which happen to be on albums, but everything is on albums isn't it?).  What am I missing that I fail to understand where you are coming from?  It's a mystery to me.

Now as it happens, I do have another tag that uniquely identifies an album, and is on every track.  However, many people won't.  So in my case 'setid' is the old-fashioned catalogue number.  Others may have industry-standard discogs references and so on.  Is there any chance this could be configurable?

**Comment by Pastim** [comment]
I've been testing the new version.  With groups that only occur on a few albums all is well.  With groups that are on, say, 100+ albums (for instance I have 118 albums with piano concertos), it starts playing after less than a minute, but is still 'exploring' many minutes later,  Is that expected?  I am currently waiting for it to finish over 8 minutes after it started playing the first one.  

I have a feeling that this may explain some of the behaviour I reported earlier - I thought it had finished searching but it hadn't, and then it got confused and 'stuck'.

What does 'cancel' do - just stop searching for more and leave the current selection in place, or what?

I don't know how this all works so can't suggest much.  Where is the search happening?  Client or server?

Sorry to be such a problematic user!

**Comment by Pastim** [comment]
The 'exploring' stage doesn't seem to finish.  The one I reported above is still going after nearly an hour.  I've stopped it now.

**Comment by medoc92** [comment]
I'm very sorry about this, I concentrated so much on testing the cancellation that I never tried to see what happened on a normal exit (except for very short cases where the dialog never appears). So I just did not delete the dialog at the end of the walk. It's fixed now. You're not a problematic user, you are a very kind one who accepts to be experimented upon, thank you for this. But this was pushing it a bit far, and I'll try to be more thorough in the future. For my excuse, it was the end of the day...

About why I need album information: the thing I think you may be missing is that there is not a lot of information in the UPnP track record. Only a very small subset of the tags is present in there, namely artist, album(replaced by group by MinimServer, title, track number. So for the same movement of the same group recorded by some artist on two separate discs. I might have:

    Mr Smith, Concerto No 1, Allegro, track 1
    Mr Smith, Concerto No 1, Allegro, track 1

Notice how they are the same :-)

This is why the groups ended sorted with all tracks one first, then twos etc. in the first version.

I'll give a name to this small bunch of info and call it "Track" in the following.

I have  an other source of information for helping with this, which is the path I followed in the tree to reach this particular Track. In the case of Minim, if the 2 paths are:

    Group/Concerto No 1/Track
    Group/Concerto No 1/Track

You will notice that it does not really help (in a pinch, I could give sequence numbers to identical entries encountered in the same directory, but this would complicate matters a lot). Now if the path has an album component:

    Album/Concerto 1,3,5 1971/Group/Concerto No 1/Track
    Album/Concerto 1,3,5 1982/Group/Concerto No 1/Track

Then I can know what goes with what. The album names are bound to be different: else it's the same recording, and there is nothing I can do. 

In the case of multiple recordings on the same physical disc, of course, the track numbers will be disjoint.

In the case of multiple recordings on the same multi-disc Album, either the actual Album title will include the disc number, or the track number will be different (if the discs were merged, does this happen?). 

About setid, discog reference etc: they are just not present, ever, in the Track, so I can't use them.
Only extra elements which appear in the MinimServer tree can be used (I sort of gathered that this was what MinimServer called "Index" elements?)

**Comment by Pastim** [comment]
Thanks very much.

I've tested the update and all now works well, even playing all the 'classical' works.

I also now understand why you refer to 'random play by albums', since some tracks won't have groups, and many servers won't support groups at all.  I hadn't fully appreciated how minimserver was subverting the album data.  I guess the accurate name should be "Random play by albums/groups" so as not to confuse other users too much!

On multi-disc albums, do you get the "discnumber" tag?  Maybe it is merged with the track number?

I wonder whether using the 'item tags' feature of minimserver will help.  If Album or discogs reference was added would this help?  Do you see any more data if tags are added to this?

**Comment by medoc92** [comment]
I guess that item tags could be used, in conjunction with "Tag formatting". We would need a value which is common to a Group, and unique among all recordings of the same Group, and then use tag formatting so that it is actually sent as part of the UPnP item (what I called Track above), probably as part of the album field (where Minim displays the group name), so this column would look like "group name -session identifier). I don't know what could be used for this, and this is all a Minim server config issue anyway, not much that upplay can do about it.

The big advantage of having actually differenciated tracks would be that I could use the MinimServer "Search" feature to gather them much more efficiently than by walking the tree to record the path information.

About multi-disc albums, I had a better look at this and the option of showing discs inside a merged album is not going to work to well if the same Work appears on multiple discs, because Minim does not merge the track numbers in this case (each work begins with track 1). So there is no way to discriminate tracks from different Work versions. If the discs are fully merged, then there is a single track list and no problem. An example follows, but I am going to stop exploring this, because this is really all about MinimServer configuration.

== Album title: "bla bla cd n"  -> No merge at all 

Discs show as separate albums, no problem


== Album title: "bla bla", using disc numbers -> full merge

(servers) > MinimServer[y] > 7 albums > Brandenburg Concertos - Italiano

Concerto No. 1	Concerto Ita...
Concerto No. 2	Concerto Ita...
Concerto No. 3	Concerto Ita...
Concerto No. 4	Concerto Ita...
Concerto No. 5	Concerto Ita...
Concerto No. 6	Concerto Ita...


(servers) > MinimServer[y] > 7 albums > Brandenburg Concertos - Italiano > Concerto No. 4

17	Concerto No. 4 in G major, BWV 1049: I .  ...
18	Concerto No. 4 in G major, BWV 1049: II.  ...
19	Concerto No. 4 in G major, BWV 1049: III. ...

Ok, because the track numbers are all unique


== Album title: "bla bla [disc n]" - Merge showing separate discs 

(servers) > MinimServer[y] > 7 albums > Brandenburg Concertos - Italiano -

_>> Disc 1
Concerto No. 1	Concerto Ita...
Concerto No. 2	Concerto Ita...
Concerto No. 3	Concerto Ita...
_>> Disc 2
Concerto No. 4	Concerto Ita...
Concerto No. 5	Concerto Ita...
Concerto No. 6	Concerto Ita...


(servers) > MinimServer[y] > 7 albums > Brandenburg Concertos - Italiano > Concerto No. 4

1	Concerto No. 4 in G major, BWV 1049: I.   ...
2	Concerto No. 4 in G major, BWV 1049: II.  ...
3	Concerto No. 4 in G major, BWV 1049: III. ...


-> Problem if the same work appears several times (duplicate track 1), we'd have to somehow add the disc info to either the track or album (group) column.

By the way, in this configuration, if you duplicate one disk, just changing the Album Title tags with [disc n+1], minim sort of gives up and does not properly display the duplicate discs.

**Comment by Pastim** [comment]
Can you briefly explain to me the source of the data you are acting on?  Is it exactly the 4 columns shown when looking at a list of individual tracks, and you are randomising on the 3rd column from the left?

As to the multi-disk album issue, I'n not sure this affects me, since I always ensure that the group name is different if there are two or more versions of a work on an album.  Is that correct?

**Comment by Pastim** [comment]
I just had a look at the track numbers displayed on each line item.  On a multi-disc work minimserver increments the numbers.  What follows is the 4 columns split onto 4 separate lines.  The Track title includes the disc number and tracknumber (because I included them in the Title format).

99	

4:05 - St Matthew Passion BWV.244 - Part 2 - 34. Tenor - Recitativo: Mein Jesus schweigt zu falschen Lügen stille (1988) [16/44100]	

Johann Sebastian Bach - John Eliot Gardiner, Monteverdi Choir, English Baroque Soloists, London Oratory Junior Choir, John Eliot Gardiner, Howard Crook	

St Matthew Passion BWV.244 - Part 2

01:15

**Comment by Pastim** [comment]
I think I have solved it for my purposes at least.

As I mentioned above it wasn't clear to me that you were using the actual display format of each track in a list to randomise on.  I should have realised this, but I had imagined some internal UPnP protocol was involved.

So now understanding this I worked out how to fix it for my own purposes. In minimserver, tagFormat, set it to:

  Group.displayFormat={$album^^: $group}

ie: prefix the group with the album.

This has a side effect of displaying the album name in the now-playing display, along with the group (as "album: group"), which suits me fine.  The same is true if you use Group as an index when you get to the item list - I'm OK with that.

It seems to have a side effect of slowing down the randomisation - but this may be my imagination.

For cases where there is more than one version on an album, this still works for me because I use different group names (and minimserver insists that a group be sequential tracks so in many cases using the same group name on one album would fail anyway). 

Any views?

**Comment by Pastim** [comment]
A comment on the speed of reandomisation (or possibly control of the dialogue).

If I pick a Genre and play random groups from there, the dialogue for even 30 tracks can stay up for a very long time.  Clicking on another genre kills the dialogue and the tracks start to play.  If the number of tracks is large (hundreds or thousands), it still seems to count the tracks very quickly, but then sticks there.

If I pick a Genre, click and select nnn Albums and then play random groups, it's still very quick.

Strange.

You have now supplied a CP that out does anything else I know of.  Brilliant.  As and when you release this I'll certainly post comments on the minimserver website, and yours of course, recommending it.

**Comment by medoc92** [comment]
The problem when trying to use the minim tree for automatic exploration is that it is extremely complicated. A given track can be found through many different paths. Actually, I am not fully certain that there are no loops in there. 

I've tried launching "play by groups" on the top Genre directory, when I stopped it a few minutes later, it had found more than 1 million duplicate tracks (thanks Intel for fast machines...). I could check if it was actually looping, but I think that there is a good chance that this is just the amplification due to the structure of the tree. 

If you pick nnn albums, the subtree under this is enormously smaller, you get very quickly to tracks, you are below all the sections for Artist, "All artists", which I think are quite hairy.

I think that it's a nice idea to include the album identifier with the Group, if it does not lead to too much displayed information. Actually this would probably allow using the Minim "search"  command, and avoid the tree walk entirely. Maybe I'll push a version which does this, just for testing how it works with your setup (hard-wiring this is a 1-line change). If it works well, I could include the choice in the preferences.

But doesn't changing the Group identifier this way forbid to easily start "play all versions of this group" ?

**Comment by Pastim** [comment]
Firstly, performance.  I pick a Genre, say 'Flute concerto', and play random groups.  There are 53 (so it says), but it takes forever unless I click on any other genre entry, then it stops the dialogue.  Conversely, I click on Flute Concerto, then on the nnn albums immediately under the genre, and it all happens very quickly.  Surely there should be the same number of items to search?  It is almost as if it is still searching the entire set of tracks when picking at the genre level.

And so to Group display format.  The key here, as I understand it, is that when using the Group as an index (in Index Format), it is shown as just the group, since that is the index.  It is only when you get further into the tree that it uses the Display Format.  So yes, I can still randomly play all versions of the group by selecting the Group index entry.

**Comment by Pastim** [comment]
With the most recent version, and with my tweak to the group display, when the selection gets large (around 3000 or so tracks I think) minimserver crashes.  I've reported it over there, but also copy it here in case you have any ideas, or if it could be something upplay is doing.

20:23:09.806 MinimServer on VB is running
20:23:49.672 MinimServer: exception while processing action request: java.lang.IndexOutOfBoundsException: Index: 0, Size: 0
java.lang.IndexOutOfBoundsException: Index: 0, Size: 0
	at java.util.ArrayList.rangeCheck(ArrayList.java:653)
	at java.util.ArrayList.get(ArrayList.java:429)
	at com.minimserver.Server$Group.getCompleteAlbum(Server.java:12399)
	at com.minimserver.Server$Group.copyValuesFirstTag(Server.java:12299)
	at com.minimserver.Server$TagValueFormat.copyValues(Server.java:9037)
	at com.minimserver.Server.addAlbumTag(Server.java:6969)
	at com.minimserver.Server.access$12700(Server.java:114)
	at com.minimserver.Server$TrackItem.getDIDL(Server.java:14400)
	at com.minimserver.Server$Container.getContentsDIDL(Server.java:14946)
	at com.minimserver.Server$ContentDirectoryProvider.doBrowse(Server.java:19214)
	at com.minimserver.Server$ContentDirectoryProvider.browse(Server.java:19165)
	at org.openhome.net.device.providers.DvProviderUpnpOrgContentDirectory1$DoBrowse.actionInvoked(DvProviderUpnpOrgContentDirectory1.java:978)
20:23:49.795 MinimServer error: now stopped, use 'restart' or 'exit'
20:24:20.779 HTTPConnection: exception while processing HTTP request: java.lang.NullPointerException
java.lang.NullPointerException
	at org.jminim.lib.HTTPContext.getHandler(HTTPContext.java:87)
	at com.minimstreamer.WaveResourceBase.writeData(WaveResourceBase.java:121)
	at com.minimserver.ServerRequestHandler.processRequest(ServerRequestHandler.java:529)
	at org.jminim.lib.HTTPConnection$WriterThread.runWriterThread(HTTPConnection.java:458)
	at org.jminim.lib.HTTPConnection$WriterThread.run(HTTPConnection.java:418)

**Comment by Pastim** [comment]
A brief update.  I found an album (by reducing the search criteria) that causes the minimserver crash when I defined the displayFormat for Group.  I checked that it crashes whatever CP I use, so, as I thought, upplay (and randomisation) has nothing to do with the problem.  I'm reporting this on the minimserver forum.

**Comment by medoc92** [comment]
In any case, upplay does not do anything "illegal" with the MinimServer interface, so any crash is bound to be a MinimServer bug. It's nice that you found a specific album to cause the crash though, it will make the resolution much easier. 

By the way it seems that duplicating one disk in an album (just changing the cd number) also is troublesome for MinimServer (no crash, but the processing of the groups appears to be disabled for the dup disks).

**Comment by Pastim** [comment]
I found the exact data that causes minimserver to crash, and can work around it.  I have, for example, an album with 6 Works all called 'Stabat Mater', which all had the same Group.  With the modified Group displayFormat I use this caused the crash (why that is I cannot say, but it definitely causes the crash).  I found 5 other albums with similar issues, and have fixed them all, so I get no more crashes.

If you look at the minimserver log it reports when most issues are found at start-up/rescan.  I had to fix several things (like missing disks on albums because I had duplicates elsewhere) when setting it up initially.

So I have two remaining queries.  The first is the performance delay if I pick something like a large Genre (or Composer etc).  As I understand it from your explanation, if I pick Genre, you have to look at all the possible tracks for each and every one of the possible index tags that are underneath this Genre, meaning you see many duplicates.  Is that correct?  The track count goes up to maximum quite quickly, but then stops and appears 'stuck', although in fact all the tracks (and groups) have been found, and cancelling returns all the groups.  So could just cancel each time the count stops increasing.  Is that correct?  Is there no way to avoid this except by my manually clicking on the Genre (or whatever) and going to the next index level below that?

Secondly, you always return just one Group (or Album) to the playlist.  Is that part of the randomisation process or have you already got a complete list and are just feeding it one at a time to the playlist?  Would it be possible to return multiples (always complete) Groups up to some limit of tracks?  Some of my Groups can have up to 60 or more tracks, so the limit would need to be fairly high.

**Comment by medoc92** [comment]
About the first question, your understanding is correct, and yes, probably cancelling when the count stabilizes should do what you want in most cases. This is all dependant of the structure of the MinimServer tree though, and there is no warranty. Maybe Simon would be able to define a better strategy for walking the tree with this purpose, but I don't know enough to do better. I'd have no issue implementing a Minim-specific strategy though, the code already knows at what server it is speaking. My guess is that the Artist and All Artists branches under any "Genre" generally explain a big part of the amplification.

About the second issue, there would be no problem feeding several groups at a time but I am not sure I understand your purpose: this would run contrary to being able to skip a single group. What would be the advantage ?

**Comment by Pastim** [comment]
Performance - yes, I suspect that's true.  The nnn Albums or xxx items would presumably be the shortest if it was possible to automate.  Given that I can cancel, or stop being lazy and do it manually myself, it's no big deal!

My idea was simply to show what's next, much like random tracks.  I can manually skip to the next group if I was.  Maybe as an option?  Again, not a big issue, just a comment :)

All is fundamentally very well indeed.  Time to let other users have the benefit?

**Comment by medoc92** [comment]
Yes, it will soon be time to release something and close this quite long-running issue...
I do think that some kind of feedback that random-play-by-group is running, and maybe a look-ahead of the next thing to play would be nice to have though, still thinking about it.

**Comment by Pastim** [comment]
Agreed, but don't wait too long. Let some others complain about some of my daft ideas :)

**Comment by medoc92** [comment]
This will be a bit of an easter egg, but I now set the next group to play description as tooltip on the "Clear Playlist" button (the broom). This is not totally without logic as clicking the broom will actually play what it shows... So closing this issue at last, I'll soon release another version, but I have a few things to do for the windows version first.

**Comment by Pastim** [comment]
Good idea - I've just tried it.

It's been a real pleasure for me to test and try to help you develop this idea of mine.  I'm sorry I'm so slow, but I usually understand in the end.

Let me know when a formal release is out & I'll tell others about it.

**Comment by Pastim** [comment]
I'm having a few problems.  Sometimes when i am waiting for the next group to play, nothing comes up.  It says it's still in Random play by group (the stop random play option is still available) but the playlist is empty,  The artwork is of the group just played.

I don't think I've sen this before today, with the helpful tooltip.

**Comment by medoc92** [comment]
I'll try to do more testing today, to confirm that the issue is linked to the tooltip changes, you can use:

    git checkout c357b81cbf4b896ecaaf529f553adbae9ac45405

to return to the state just before them.

**Comment by Pastim** [comment]
I'm mainly using a Musical Fidelity M1 CLiC.  The problem is intermittent (the worst sort).  After I reported I tried again (restarting upplay) and it did it again after the first track.  I then tried some other random groups without stopping upplay and it worked OK.  I haven't yet seen a pattern, except possibly being more likely to happen when the group is large (many hundreds if not thousands of tracks).  I will see if I can test some more later today.

**Comment by medoc92** [comment]
Don't lose too much time on this, I think that the method I currently use to detect the end of the current playlist is unreliable (e.g. it does not work with the Bubble renderer, which is surprisingly crappy by the way). Unfortunately, implementing robust logic for this is not completely trivial (there are many cases to consider, taking into account user and other CP actions). I'm hopeful I'll find a way, but it will take a little time. It would still be interesting to know if it is linked with the tooltip changes, but I'd actually be surprised (except if there is a subtle timing dependancy, which I would need to fix anyway).

**Comment by medoc92** [comment]
I've pushed a change which should hopefully more reliably detect the end of the playlist. Seems to work on both upmpdcli and bubble. By the way, is the Musical Fidelity an openhome renderer ? Else, the changes won't affect it, and I'll got back to see what's wrong with the upnp av mode.

**Comment by medoc92** [comment]
oops changes really there now (they were not on my last message).

**Comment by Pastim** [comment]
I was using it with upmpdcli earlier and all seemed OK, but I have previously not used upmpdcli that much.

I'm not even sure what an openhome renderer is to be honest, but I don't think the MF is.

Unfortunately minimserver has just been updated but has a problem on my system, so I may not be able to test for some time, possibly days.

**Comment by Pastim** [comment]
I just got round to playing the latest version on my MF M1 CLiC.  The first Group was quite long, so I skipped some tracks and eventually used the brush.  All then got very confused. It showed a new Group in the playlist, but didn't play until I hit the play button.  Then the track being played was the first on the list, and shown on the MF, but the title at the top of the playlist was a corrupt form of a track from the previous Group.  There was also a stray track from the first Group at the end of the playlist.

Some time later, after a minimserver restart, it's behaving almost perfectly, but I am sometimes getting a stray track from the previous Group left on the list.

**Comment by Pastim** [comment]
I have a feeling the issue may not be directly to do with random play.  I think that upplay is moving to the next list, reading the current track from the renderer before it's actually finished playing (or maybe the renderer is slow in clearing out info about the last track), and adding that to the new playlist, sometimes getting confused in the process.  One reason I think that is the title of the entry from the old Group being adding to the new Group is often slightly corrupted, and that sounds familiar from previous problems with upplay was using the MF's version of a title.

Does that sound plausible?  It's hard to emulate with any other way of playing other than automatic playlist generation.  It is certainly intermittent, and I haven't seen it (yet!) with upmpdcli.

I suggest closing this again and if I get more issues I'll open a new issue.

**Comment by Pastim** [comment]
Sorry, me again.....

If I play random groups of which there are very few groups, it seems to play one less than the full set.  When playing the penultimate one the tooltip shows one more to go, but it doesn't play.  Thus I have 2 bassoon concertos.  The tooltip confirms this, but I only ever get to play one.

'Random' also produces results one doesn't expect.  I have played random Piano Solos several times, and some of the larger groups (ie those with many tracks) seem to have come up more often than I'd expect,  I'm sure that it's the way of randomness, but can you confirm upplay picks random groups and playing them, not random tracks and then playing the groups they are in?

**Comment by Pastim** [comment]
Apart from the last comments I made above, the version I pulled and compiled a few hours ago seems very stable on both upmpdcli and my MF M1 CLiC,  

Minimserver has also been updated for a few issues so I have been able to test freely for the last few hours.

**Comment by medoc92** [comment]
About the issues with the M1 click playlist mixups, I giess that a mixup with data from the renderer is possible, but difficult to diagnose without the renderer. If this persists in an ennoying way, I will make sure that the software logs the events we would be interested in, and you will record a log so that I can try and get an idea of what is happening.

I fixed a bug where upplay would never play the last group in the random group list.

About bigger groups being selected with more probability, you are right, upplay selects a track, and then plays the group it's in, this was just easier. Kudos for guessing what was happening. I'm going to change the selection method.

**Comment by Pastim** [comment]
The 'last group' bug has gone, and after a couple of quick trials I'm getting a better set of random groups now, rather than a concentration of groups with many tracks.

Thanks for all that, and for implementing the album art idea.  Very helpful.

Lats night the M1 was behaving pretty well, so I'll just keep an eye on it.  If it carries on too much I'll let you know and try to produce logs to help.  Are there any logs I should be looking at now?

**Comment by medoc92** [comment]
I'm afraid that the most significant change for group selection was not in there yet. You may be seeing the effect of changing the method for seeding the random sequence. Since the change I just pushed, all groups should have the same chance, independantly of size.

About the logs, don't bother. When we want to look at this, I'll prepare the code to output more precisely the interesting data.

**Comment by Pastim** [comment]
That should teach me to remember random doesn't mean predictable (as I said in my original comment)!  At the time a couple of quick tests seemed better. Anyhow, I'm now up to date on my 2 main controlling systems (laptop in lounge and desktop in study).

So I started playing, and all of sudden lost sound.  I checked all the settings, server, upplay and scratched my head.  Restarted minimserver.  Rebooted the server. Restarted the laptop.  Nothing.  After about 15 minutes I saw the power amp light had gone out. Doh.  An electronics failure for once.  Some quick tests, plugs, fuses, etc.  Nothing.  Internal fuses, power switch on amp, all OK. Out with the ladder and up into the loft where the old stuff is still kept (resisting all attempts to be thrown away).  So I'm live again, using a 1979 power amp that still works.  All this modern (and expensive) stuff is so fragile!

Sorry, just a tale of a tester.  

So I'm testing again on the latest upplay, and all's well. With piano solos I certainly seem to be getting more single tracks, although that could just be randomness or my imagination.   

I think the problems I sometimes get on my MF M1 CLiC are when I am skipping tracks quite quickly to do some tests.  If I just let it play it seems stable, so I don't want to pursue this for now at least.

Thanks again for all your extraordinary quick and successful work on this.  I hope and trust others will appreciate it.

**Comment by medoc92** [comment]
Thanks for helping putting this together, it started from quite far  from being usable... 

I'm releasing version 1.1 with the current code, there will be minor releases if we find other bugs.

I've had the same small cheap Rega Elex amp for maybe 25 years, it has completely destroyed my  speakers *twice* (in the careful hands of my children and their friends), but it is still going strong... I see that the current model is about the same price in Euros that what I paid in Francs at the time (1 Euro = 6.56 francs or such). I'm not sure we're making progress...
